### PR TITLE
Add `SemanticVersion.valid?` & `SemanticVersion.parse?`

### LINKED ISSUE #15020: Add `SemanticVersion.valid?` and `SemanticVersion.parse?`
Currently the only way to "test" a semantic version string is to rescue from `SemanticVersion.parse`. Such helper methods would be really convenient given that there's no regex or alternative method publicly exposed to do so. It seems like these once existed in an old PR reworking the API and just didn't make it to the existing one, and looking at the implementation it's relatively straightforward to add.

### PR DESCRIPTION
**Pull request by devnote-dev**
I wasn't sure if this required additional specs given that the underlying implementation (i.e. the regex) is the same. Closes #15020.

### CODE REVIEW

File: src/semantic_version.cr [hunk 1]
```diff
@@ -30,18 +45,23 @@ struct SemanticVersion
...
-    end
+    parse?(str) || raise ArgumentError.new("Not a semantic version: #{str.inspect}")
+  end
+
+  # Parses a `SemanticVersion` from the given semantic version string.
+  #
+  # ```
+  # require "semantic_version"
+  #
+  # semver = SemanticVersion.parse("2.61.4")
```
Comment (Sija) [inline src/semantic_version.cr:56]:
```suggestion
  # semver = SemanticVersion.parse?("2.61.4")
```

**Review by Sija — CHANGES_REQUESTED** [review]

File: src/semantic_version.cr [hunk 2]
```diff
@@ -19,7 +23,18 @@ struct SemanticVersion
...
 
-  # Parses a `SemanticVersion` from the given semantic version string
+  # Checks if *str* is a valid semantic version.
+  #
+  # ```
+  # require "semantic_version"
+  #
+  # SemanticVersion.valid?("1.15.0") # => true
+  # SemanticVersion.valid?("1.2")    # => false
+  def self.valid?(str : String) : Bool
```
Comment (oprypin) [inline src/semantic_version.cr:33]: Do we need to add the method `valid?` if `parse?` can be used in exactly the same way? And I think currently there isn't any significant performance benefit to it either

Comment (devnote-dev) [reply src/semantic_version.cr:33]: I think it's more semantics, for example, `Hash#has_key?` vs `Hash#[]?`: they can be used in the same way and have the same underlying implementation, but the methods themselves convey different things. Comparing performance feels like splitting hairs, I agree that it's likely minimal.

Comment (straight-shoota) [reply src/semantic_version.cr:33]: I think this is a valid point. I would read the semantic meaning of `parse?` as "can it parse this?". If not, the return value is falsey. If yes, it's the parse result.

A differentiation into a specific bool method like `valid?` is particularly relevant when validation can be implemented more efficiently than parsing.

**Review by nobodywasishere — APPROVED** [review]

**Review by straight-shoota — APPROVED** [review]

**Review by HertzDevil — APPROVED** [review]

**Review by straight-shoota — CHANGES_REQUESTED** [review]
Actually, we should still have some basic specs to ensure the methods work.

**Comment by devnote-dev** [comment]
@straight-shoota can you elaborate on what specifically? All the methods depend on the regex which is already tested in specs.

**Comment by straight-shoota** [comment]
Just a few simple calls to `.valid?` and `.parse?` with a successful and a failing value for each should be fine.
This is to make sure the methods compile and operate generally. Edge cases are covered by the other implementations, but we still need to validate these particular API methods are working.

**Comment by straight-shoota** [comment]
Can we get at least one successful test for each method? At the moment, even `false`/`nil` would fulfil the specs.

```cr
SemanticVersion.valid?("1.2.3").should be_true
SemanticVersion.parse?("1.2.3").should eq SemanticVersion.new(1, 2, 3)
```

**Comment by beta-ziliani** [comment]
Hi @devnote-dev , thanks a lot for your contributions! We understand your reaction in [this comment](https://github.com/crystal-lang/crystal/pull/15051#issuecomment-2458081940); having several rounds of reviews is tiring! Let me take a bit more of your time to understand what are we doing wrong.

From our point of view, the comment [suggesting specs](https://github.com/crystal-lang/crystal/pull/15051#issuecomment-2429032852) was clear enough: "add successful and failing cases". Since only half of that was added in the subsequent commit, straight-shoota naturally asked for the other half. This seemed to not be excepcted, and therefore the reaction.

First of all, and we need to make this more obvious to the contributors: whenever you're done with a PR, if you don't mind someone else taking over, you can leave it to us to do the final polish. And second, in order to improve future reviews, could you elaborate on why the first comment failed to transmit the expected outcome?

Thanks once again for making Crystal better!

**Comment by devnote-dev** [comment]
@beta-ziliani Thank you for your acknowledgement. I don't mind having several rounds of reviews, but I do find these recent ones a bit unnecessary.

I understand and support having full test coverage, but I also believe there's a point where testing is overdone. The `.parse` method's implementation depends on `parse?`, so IMO it only really makes sense to test that it raises the correct exception. We know that it will parse correct versions as long as `.parse?` does. I have similar feelings for testing `.valid?`: for as long as `.parse?` will parse correct versions, `.valid?` will return `true` for them. The API is very closely intertwined which is why I reacted as I did, it was nothing personal towards @straight-shoota but rather the reasoning for the review (sorry if it came across that way 🙏).

I won't protest this issue as I've seen it become a blocker for valid contributions in the past, but I would like this issue to be taken into consideration for future contributions.

**Review by straight-shoota — APPROVED** [review]

**Comment by ysbaddaden** [comment]
@devnote-dev I agree it can feel repetitive or be over-testing, but such assumptions means that specs become implementation dependent. What if `#parse` is refactored and doesn't depend on `#parse?` anymore, same for `#valid?`. Specs should catch any issues but they won't, and they fail at their job to protect against regressions and help with refactors.

**Review by ysbaddaden — APPROVED** [review]
